src/Symfony/Component/PropertyAccess/Mapping/Factory/LazyLoadingMetadataFactory.php

```diff
@@ -0,0 +1,166 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\PropertyAccess\Mapping\Factory;
+
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\PropertyAccess\Exception\NoSuchMetadataException;
+use Symfony\Component\PropertyAccess\Mapping\ClassMetadata;
+use Symfony\Component\PropertyAccess\Mapping\Loader\LoaderInterface;
+
+/**
+ * Creates new {@link ClassMetadataInterface} instances.
+ *
+ * Whenever {@link getMetadataFor()} is called for the first time with a given
+ * class name or object of that class, a new metadata instance is created and
+ * returned. On subsequent requests for the same class, the same metadata
+ * instance will be returned.
+ *
+ * You can optionally pass a {@link LoaderInterface} instance to the constructor.
+ * Whenever a new metadata instance is created, it is passed to the loader,
+ * which can configure the metadata based on configuration loaded from the
+ * filesystem or a database. If you want to use multiple loaders, wrap them in a
+ * {@link LoaderChain}.
+ *
+ * You can also optionally pass a {@link CacheInterface} instance to the
+ * constructor. This cache will be used for persisting the generated metadata
+ * between multiple PHP requests.
+ *
+ * @author Luis Ramón López <lrlopez@gmail.com>
+ */
+class LazyLoadingMetadataFactory implements MetadataFactoryInterface
+{
+    /**
+     * The loader for loading the class metadata.
+     *
+     * @var LoaderInterface|null
+     */
+    private $loader;
+
+    /**
+     * The cache for caching class metadata.
+     *
+     * @var CacheItemPoolInterface|null
+     */
+    private $cache;
+
+    /**
+     * The loaded metadata, indexed by class name.
+     *
+     * @var ClassMetadata[]
+     */
+    private $loadedClasses = array();
+
+    /**
+     * Creates a new metadata factory.
+     *
+     * @param LoaderInterface|null        $loader The loader for configuring new metadata
+     * @param CacheItemPoolInterface|null $cache  The PSR-6 cache for persisting metadata
+     *                                            between multiple PHP requests
+     */
+    public function __construct(LoaderInterface $loader = null, CacheItemPoolInterface $cache = null)
```

Comment:
I just replicated the `Validator` way of accessing metadata to ease a future refactoring (i.e. when Symfony gets its own `Metadata` component) so I wouldn't like to diverge too much from the established implementations.

@dunglas, Do you still want me to implement the cache as a decorator?